keras_cv/layers/vit_layers.py

```diff
@@ -0,0 +1,243 @@
+class Patching(layers.Layer):
+    """
+    Layer to turn images into a flat sequence of patches for Vision Transformers.
+
+    Reference:
+        An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale
+        by Alexey Dosovitskiy et al. (https://arxiv.org/abs/2010.11929)
+
+    Based on Khalid Salama's implementation for:
+        https://github.com/keras-team/keras-io/blob/master/examples/vision/image_classification_with_vision_transformer.py
+
+    The layer expects a batch of input images and returns batches of patches. If the height and width of the image
+    isn't divisible by the patch size, the supplied padding type is used (or 'VALID' by default).
+
+    Args:
+        patch_size: the size (patch_size, patch_size) of each patch created from the image
+        padding: default 'VALID', the padding to apply when extracting patches from images when the shape isn't
+            divisible by the patch_size
+    Returns:
+        batch of "patchified" and then flattened input images.
+
+    Basic usage:
+
+    ```
+    img = url_to_array(url)
+    batch_img = np.expand_dims(img, 0)
+
+    patch_size = 16
+    patches = keras_cv.layers.Patching(patch_size)(batch_img)
+    print(f"Image size: {batch_img.shape}")                # Image size: (1, 224, 224, 3)
+    print(f"Patch size: {(patch_size, patch_size)}") # Patch size: (16, 16)
+    print(f"Patches: {patches.shape[1]}")            # Patches: 196
+    ```
+    """
+
+    def __init__(self, patch_size, padding, **kwargs):
+        super().__init__(**kwargs)
+        self.patch_size = patch_size
+        self.padding = padding
+
+    def call(self, images):
+        """Calls the Patching layer on a batch of images.
+        Args:
+            images: A `tf.Tensor` of shape [batch, height, width, channels]
+
+        Returns:
+            `A tf.Tensor` of shape [batch, patch_num, patch_dims]
+        """
+        batch_size = tf.shape(images)[0]
+        patches = tf.image.extract_patches(
+            images=images,
+            sizes=[1, self.patch_size, self.patch_size, 1],
+            strides=[1, self.patch_size, self.patch_size, 1],
+            rates=[1, 1, 1, 1],
+            padding=self.padding,
+        )
+        if self.padding not in ["SAME", "VALID"]:
+            raise ValueError(
+                f"Padding must be either 'SAME' or 'VALID', but {self.padding} was passed."
+            )
+
+        patch_dims = patches.shape[-1]
+        patches = tf.reshape(
+            patches, [batch_size, patches.shape[-2] * patches.shape[-2], patch_dims]
+        )
+        return patches
+
+    def get_config(self):
+        config = {
+            "patch_size": self.patch_size,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class PatchEmbedding(layers.Layer):
```

Comment:
can we put patching and patch_embedding in their own .py files?

Reply:
They originally were, but were put into a single file after: https://github.com/keras-team/keras-cv/pull/929#issuecomment-1306515688

Reply:
I'd personally opt for having them be standalone layers in different files, since they'll be used/reused for other vision transformers, but whatever is the team ruling is what I'll go with ;)

Reply:
Do we keep them in the same file or do we separate them in the end? @LukeWood